pytato/target/pycuda/__init__.py

```diff
@@ -0,0 +1,188 @@
+        assert callable(variables_after_execution[self.entrypoint])
+        return variables_after_execution[  # type: ignore[no-any-return]
+            self.entrypoint]
+    
+    @cached_property
+    def _bound_argment_names(self) -> Set[str]:
+        return set(self.bound_arguments.keys())
+    
+    @cached_property
+    def _processed_bound_args(self):
+        import pycuda.autoinit
+        import pycuda.gpuarray as gpuarray
+        import numpy as np
+        processed_bound_args = {}
+        for name, arg in zip(self.bound_arguments.keys(), self.bound_arguments.values()):
+            if isinstance(arg, np.ndarray):
+                processed_bound_args[name] = gpuarray.to_gpu(arg)
+            elif isinstance(arg, gpuarray.GPUArray):
+                processed_bound_args[name] = arg
+            else:
+                raise ValueError("Array format not supported")
+        return processed_bound_args
+
+    def __call__(self,  *args: Any, **kwargs: Any) -> Any:
+
+        if args:
+            raise ValueError(f"'{type(self).__call__}' does not take positional"
+                             " arguments.")
+
+        if set(kwargs.keys()) & self._bound_argment_names:
+            raise ValueError("Got arguments that were previously bound: "
+                    f"'{set(kwargs.keys()) & set(self.bound_arguments.keys())}'.")
+
+        updated_kwargs = dict(self._processed_bound_args)
+        updated_kwargs.update(kwargs)
+        updated_kwargs = {kw: arg
+                          for kw, arg in updated_kwargs.items()
+                          if kw in self.expected_arguments}
+        return self._compiled_function(**updated_kwargs)
+
+    def copy(self, **kwargs: Any) -> BoundCUDAGraphProgram:
+        from dataclasses import replace
+        return replace(self, **kwargs)
+
+    def with_transformed_program(self, *args: Any, **kwargs: Any
+                                 ) -> BoundCUDAGraphProgram:
+        raise ValueError("Cannot transform python program.")
+
+# }}}
+
+
+# {{{ numpy-like python targets
+
+class NumpyLikePythonTarget(Target, ABC):
+
+    @abstractproperty
+    def numpy_like_module_name(self) -> str:
+        pass
+
+    @abstractproperty
+    def numpy_like_module_name_shorthand(self) -> str:
+        pass
+
+    @abstractmethod
+    def bind_program(self,
+                     program: str,
+                     entrypoint: str,
+                     expected_arguments: FrozenSet[str],
+                     bound_arguments: Mapping[str, Any]) -> BoundCUDAGraphProgram:
+        pass
+
+# }}}
+
+
+# {{{ cudagraph-numpy target
+
+
+class CUDAGraphTarget(NumpyLikePythonTarget):
+    """
+    A target that generates code for a python program by offloading array
+    operations to :mod:`cudagraph.cudagraph`.
+    """
+
+    @property
+    def numpy_like_module_name(self) -> str:
+        return "pycuda.driver"
+
+    @property
+    def numpy_like_module_name_shorthand(self) -> str:
+        return "drv"
```

Comment:
We shouldn't derive this from `NumpyLikePythonTarget`. Just PythonTarget is probably more suitable.